Charger Crusher Return

PhantomEpicness opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
Due to github fuckery, I was forced to make a new PR

old PR: https://github.com/cmss13-devs/cmss13/pull/14
## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->


Old momentum crusher has returned in form of the "Charger" crusher. The basis is simple, hold move key and destroy shit.


Passive: Directional armor. Charger has 50 frontal armor, 35 side armor, and 25 rear armor. Because of this ability, charger crusher is unable to direction lock itself.

Active: Charge
Enables charging in which you will become a living train and ram shit at the cost of being plasma heavy. Upon reaching maximum momentum, you will roar and deal full damage.

Active: Tumble
 Clicking to the side of where the crusher is facing will launch it 2 tiles to the side in the clicked direction. Hitting a human will knock them over. This ability can be used while charge to quickly offset yourself. However, it will reset your momentum to 0.

Active: Ram

An ability that allows you to fling a person in front of you for a short distance. Has a 2 second stun and a longer 10 second cooldown. Subject to change and/or removal.

Active: Charger stomp. Unlike the base crusher, it does not stun, but does heavy damage to any human unfortunate enough to be under the charger, it also applies a mild screen shake to any humans in view and can be used to help escape.


Credit to original charger devs, lumdor and donk!

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->
design doc: https://github.com/cmss13-devs/cmss13-design-docs/pull/9

_tl:dr_ 
Charger breathes new life into normal crusher by bringing back its most requested features pre-moba features in form of a strain. The caste design is much more simple compared to other xeno castes. One simple, yet tried and true fun combo exists, charging.
## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl: TotalEpicness, TheDonkified and Lumdor
add: Added charger strain for crusher.
experiment: Charger functions very similarly to old crusher. Hold movement key and after a while you gain momentum to ram humans and certain objects. Behavior upon collision with different things are subject to change.
Passive: Directional armor. Charger has 50 frontal armor, 35 side armor, and 25 rear armor. Because of this ability, charger crusher is unable to direction lock itself.
experiment: Active 1: Toggle Charging. Use this ability to toggle charging on and off.
experiment: Active 2: Tumble. Clicking to the side of where the crusher is facing will launch it 2 tiles to the side in the clicked direction. Hitting a human will knock them over. This ability can be used while charge to quickly offset yourself. However, it will reset your momentum to 0.
add: Tumble now has cool sound effects and a spin.
sprite: Added tumble ability sprites
tweak: Charger now takes 3 full hits to destroy a normal wall instead of destroying it in pretty much one. Xeno walls take half damage
experiment: Active 3: Charger stomp. Unlike the base crusher, it does not stun, but does heavy damage to any human unfortunate enough to be under the charger, it also applies a mild screen shake to any humans in view and can be used to help escape.
experiment: Active 4: Ram. A simple offensive ability that flings mobs without stun. Much like warrior fling.
add: Xenos blocking a charger's path when charger get pushed to the side, if no clear path to move to exists, they get flicker stunned so that charger may charge.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

PhantomEpicness commented:
Charger changelog
- return of base Crusher
- bonus charging armor removal
- Ram is now a 2 tile ranged warrior fling with a 2 second stun. No more mini base charge.
- fixed macro bugs, if you think macros should be rebinded, let me know
- slight health nerf and frontal armor bonus is now 25 
I am also currently working on a small shield bonus that your recieve when hitting max momentum that lasts a few seconds. This is a solution that will give charger a bit more survivability and slightly lower the risk of charging without making it excessively tanky like it was previously.
this won't make it to this TM however, as I need to see if it's even needed.

novaepee commented:
I can ensure you that all marines shall hate Ram. Over at TGMC, crusher has a similar ability, and good crusher players fish new marine players out of barricades.

Chainsawicus commented:
> A small shield bonus that you recieve when hitting max momentum that lasts a few seconds. This solution will give charger a bit more survivability and slightly lower the risk of charging without making it excessively tanky like it was previously. this won't make it to this TM however, as I need to see if it's even required.

No more shields... please...

In seriousness, since shields as a whole are being toned down or flat out removed in certain cases soon, I think you should look to alternative methods of survivability, as there's no guarantee that charger would be able to keep its shields.

PhantomEpicness commented:
> > A small shield bonus that you recieve when hitting max momentum that lasts a few seconds. This solution will give charger a bit more survivability and slightly lower the risk of charging without making it excessively tanky like it was previously. this won't make it to this TM however, as I need to see if it's even required.
> 
> No more shields... please...
> 
> In seriousness, since shields as a whole are being toned down or flat out removed in certain cases soon, I think you should look to alternative methods of survivability, as there's no guarantee that charger would be able to keep its shields.

Yeah now that it was brought up to me that shields were being done away with, I have considered other options.

stanalbatross commented:
![image](https://user-images.githubusercontent.com/66756236/179242021-2a90f18e-5608-4072-9a48-4ee50dab7feb.png)
fix

stanalbatross commented on code/modules/mob/living/carbon/human/human_damage.dm:
```diff
@@ -307,24 +310,58 @@ In most cases it makes more sense to use apply_damage() instead! And make sure t
+/*
+/mob/living/carbon/human/apply_armoured_damage(var/damage = 0, var/armour_type = ARMOR_MELEE, var/damage_type = BRUTE, var/def_zone = null, var/penetration = 0, var/armour_break_pr_pen = 0, var/armour_break_flat = 0)
+	if(damage <= 0)
+		return ..(damage, armour_type, damage_type, def_zone)
 
-		parts -= picked
-	updatehealth()
-	if(update)	UpdateDamageIcon()
+	var/obj/limb/target_limb = null
+	if(def_zone)
+		target_limb = get_limb(check_zone(def_zone))
+	else
+		target_limb = get_limb(check_zone(rand_zone()))
+	if(isnull(target_limb))
+		return FALSE
 
+	var/armor = getarmor(target_limb, armour_type)
+
+	var/armour_config = GLOB.marine_ranged
+	if(armour_type == ARMOR_MELEE)
+		armour_config = GLOB.marine_melee
+
+	var/modified_damage = armor_damage_reduction(armour_config, damage, armor, penetration, 0, 0)
+	apply_damage(modified_damage, damage_type, target_limb)
+
+	return modified_damage
+
+
```
remove commented out code all over the pr once you're done with the final implementation

PhantomEpicness commented:
Changes:

- frenzy NO LONGER affects charger at all. 
- fixed macros, they should be assigned to correct keys in order now
- 5 less back armor. Note that frontal and side armor has been changed to compensate for 5 less base armor, effectively only lowering back armor.
- hp increase 760 > 780
- fling stun removed
- fling distance 2 > 3
- changed description to a more accurate one:

> Your charge is now momentum based, the further you go, the more damage and speed you will get until you achieve max momentum when you roar. Your armor is now directional, being the toughest on the front, weaker on the sides and weakest at the back. Your shield is also removed. In return you gain more health, gain an ability to tumble to avoid enemies, and gain an ability to forcefully move enemies via ramming. Finally, you trade being able to resist slowdowns from tall autospitters for being unaffected by frenzy pheros.

stanalbatross commented:
fine in its current implementation just needs code review

Carnage068 commented:
let's gooooo

stanalbatross commented on code/game/objects/structures/crates_lockers/closets.dm:
```diff
@@ -154,6 +154,15 @@
+
+/obj/structure/closet/proc/take_damage(damage)
+	health = max(health - damage, 0)
+	if(health == 0)
```
```suggestion
	if(health <= 0)
```
sanity

stanalbatross commented on code/game/objects/structures/girders.dm:
```diff
@@ -310,11 +310,18 @@
 
+/obj/structure/girder/proc/take_damage(damage)
+	health = max(health - damage, 0)
+	if(health == 0)
```
```suggestion
	if(health <= 0)
```

stanalbatross commented on code/modules/mob/living/carbon/human/human_damage.dm:
```diff
@@ -307,24 +310,58 @@ In most cases it makes more sense to use apply_damage() instead! And make sure t
+/*
+/mob/living/carbon/human/apply_armoured_damage(var/damage = 0, var/armour_type = ARMOR_MELEE, var/damage_type = BRUTE, var/def_zone = null, var/penetration = 0, var/armour_break_pr_pen = 0, var/armour_break_flat = 0)
+	if(damage <= 0)
+		return ..(damage, armour_type, damage_type, def_zone)
 
-		parts -= picked
-	updatehealth()
-	if(update)	UpdateDamageIcon()
+	var/obj/limb/target_limb = null
+	if(def_zone)
+		target_limb = get_limb(check_zone(def_zone))
+	else
+		target_limb = get_limb(check_zone(rand_zone()))
+	if(isnull(target_limb))
+		return FALSE
 
+	var/armor = getarmor(target_limb, armour_type)
+
+	var/armour_config = GLOB.marine_ranged
+	if(armour_type == ARMOR_MELEE)
+		armour_config = GLOB.marine_melee
+
+	var/modified_damage = armor_damage_reduction(armour_config, damage, armor, penetration, 0, 0)
+	apply_damage(modified_damage, damage_type, target_limb)
+
+	return modified_damage
+
+
+*/
```
remove

PhantomEpicness replied:
done

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/abilities/crusher/crusher_abilities.dm:
```diff
@@ -51,14 +48,221 @@
+	plasma_cost = 25
+	xeno_cooldown = 10 SECONDS
+
+/datum/action/xeno_action/activable/tumble/proc/handle_mob_collision(mob/living/carbon/human/Human)
```
need a sanity check to see if the arg being parsed actually is a human

stanalbatross replied:
do

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/abilities/crusher/crusher_powers.dm:
```diff
@@ -107,7 +107,58 @@
 	return
 
-/datum/action/xeno_action/onclick/crusher_shield/use_ability(atom/A)
+/datum/action/xeno_action/onclick/crusher_stomp/charger/use_ability(atom/Target)
```
needs a target sanity check

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/abilities/crusher/crusher_powers.dm:
```diff
@@ -107,7 +107,58 @@
+			continue
+
+		new effect_type_base(Human, Xeno, , , get_xeno_stun_duration(Human, effect_duration))
+		to_chat(Human, SPAN_XENOHIGHDANGER("You are BRUTALLY crushed an stompted on by [Xeno] !!!"))
```
```suggestion
		to_chat(Human, SPAN_XENOHIGHDANGER("You are BRUTALLY crushed and stomped on by [Xeno]!!!"))
```

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/abilities/crusher/crusher_powers.dm:
```diff
@@ -107,7 +107,58 @@
+	for (var/mob/living/carbon/Human in orange(distance, get_turf(Xeno))) // MOBS AROUND
+		if (Human.stat == DEAD || Xeno.can_not_harm(Human))
+			continue
+		//new effect_type_base(H, X, , , get_xeno_stun_duration(H, effect_duration))
+		//if(H.mob_size < MOB_SIZE_BIG)
+			//H.KnockDown(get_xeno_stun_duration(H, 0.2))
+		if(Human.client)
+			shake_camera(Human, 2, 2)
+		//to_chat(H, SPAN_XENOHIGHDANGER("You are slowed as [X] knocks you off balance!"))
```
zoink this commented stuff

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/mutators/strains/crusher/charger.dm:
```diff
@@ -0,0 +1,581 @@
+	C.health_modifier += XENO_HEALTH_MOD_LARGE
+	C.speed_modifier += XENO_SPEED_FASTMOD_TIER_3
+	C.armor_modifier -= XENO_ARMOR_MOD_SMALL
+	C.ignore_aura = "frenzy" // no funny crushers going 7 morbillion kilometers per second
```
make this known to the crushers

PhantomEpicness replied:
stated in strain desc

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/mutators/strains/crusher/charger.dm:
```diff
@@ -0,0 +1,581 @@
+		CCA.stop_momentum()
+		return
+
+	health -= CCA.momentum * 40 //Usually knocks it down.
```
use a define

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/mutators/strains/crusher/charger.dm:
```diff
@@ -0,0 +1,581 @@
+		CCA.stop_momentum()
+		return
+
+	health -= CCA.momentum * 40 //Usually knocks it down.
```
as above

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/mutators/strains/crusher/charger.dm:
```diff
@@ -0,0 +1,581 @@
+	healthcheck()
+
+	if(QDELETED(src))
+		CCA.lose_momentum(1) //Lose one turf worth of speed
```
use defines for these loss amounts

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/mutators/strains/crusher/charger.dm:
```diff
@@ -0,0 +1,581 @@
+			CCA.stop_momentum()
+			return
+		X.visible_message(
+			SPAN_DANGER("[X] smashes straight into [src]!"),
+			SPAN_XENODANGER("You smash straight into [src]!")
```
```suggestion
			SPAN_DANGER("[X] smashes straight into \the [src]!"),
			SPAN_XENODANGER("You smash straight into \the [src]!")
```

stanalbatross replied:
missed

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/mutators/strains/crusher/charger.dm:
```diff
@@ -0,0 +1,581 @@
+/obj/structure/machinery/cm_vending/handle_charge_collision(mob/living/carbon/Xenomorph/X, datum/action/xeno_action/onclick/charger_charge/CCA)
+	if(CCA.momentum >= 3)
+		X.visible_message(
+			SPAN_DANGER("[X] smashes straight into [src]!"),
+			SPAN_XENODANGER("You smash straight into [src]!")
```
\the

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/mutators/strains/crusher/charger.dm:
```diff
@@ -0,0 +1,581 @@
+/obj/structure/barricade/handle_charge_collision(mob/living/carbon/Xenomorph/X, datum/action/xeno_action/onclick/charger_charge/CCA)
+	if(CCA.momentum)
+		visible_message(
+			SPAN_DANGER("[X] rams into [src] and skids to a halt!"),
+			SPAN_XENOWARNING("You ram into [src] and skid to a halt!")
```
\the

stanalbatross replied:
missed

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/mutators/strains/crusher/charger.dm:
```diff
@@ -0,0 +1,581 @@
+	var/datum/effect_system/spark_spread/s = new
+	s.set_up(5, 1, loc)
+	X.visible_message(
+		SPAN_DANGER("[X] rams [src]!"),
+		SPAN_XENODANGER("You ram [src]!")
```
\the

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/mutators/strains/crusher/charger.dm:
```diff
@@ -0,0 +1,581 @@
+		s.set_up(1, 1, loc)
+		s.start()
+		X.visible_message(
+			SPAN_DANGER("[X] rams [src]!"),
+			SPAN_XENODANGER("You ram [src]!")
```
\the

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/mutators/strains/crusher/charger.dm:
```diff
@@ -0,0 +1,581 @@
+		playsound(src, "sound/effects/metal_crash.ogg", 25, TRUE)
+		if(istype(src,/obj/structure/machinery/m56d_hmg/auto)) // we don't want to charge it to the point of downgrading it (:
+			var/obj/item/device/m2c_gun/HMG = new(src.loc)
+			to_chat(world, "TEST STEST")
```
💀

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/mutators/strains/crusher/charger.dm:
```diff
@@ -0,0 +1,581 @@
+		CCA.stop_momentum()
+		return
+	X.visible_message(
+		SPAN_DANGER("[X] rams [src]!"),
+		SPAN_XENODANGER("You ram [src]!")
```
\the

stanalbatross commented on code/modules/mob/living/carbon/xenomorph/mutators/strains/crusher/charger.dm:
```diff
@@ -0,0 +1,581 @@
+		CCA.stop_momentum()
+		return
+	X.visible_message(
+		SPAN_DANGER("[X] rams [src]!"),
+		SPAN_XENODANGER("You ram [src]!")
```
\the

stanalbatross commented:
fix conflicts + do suggestions

Carnage068 commented:
we back in this

stanalbatross commented on code/modules/defenses/bell_tower.dm:
```diff
@@ -139,7 +142,8 @@
 	var/mob/living/to_apply = target
-
+	if(HAS_TRAIT(to_apply, TRAIT_CHARGING))
+		to_chat(to_apply, SPAN_WARNING("You ignore some weird noises as you charge."))
```
return under here

PhantomEpicness commented:
Issue:

Alacrit fruit makes crusher go frenzy speed

Misttyy commented:
I really like charger except for one thing.

I've noticed a bit, and talked with other marine players to confirm it wasn't a one off thing. One charger ram can give someone 7-8 bleeds, or make them bleed across their entire body. This feels very excessive.

stanalbatross reviewed (approved):
any further issues with alacrit can be looked at with balance to the fruit itself

PhantomEpicness commented:
Made alacrit fruit work 

> I really like charger except for one thing.
> 
> I've noticed a bit, and talked with other marine players to confirm it wasn't a one off thing. One charger ram can give someone 7-8 bleeds, or make them bleed across their entire body. This feels very excessive.

It does overall damage instead of targeted damage. This prevents getting organ damage/fracs on one charge or stupid insta delimbs.

Unless you want the latter, this is how it's going to be.

Carnage068 commented:
> Made alacrit fruit work on charger, but it wanes when charging


THANK YOU.

PhantomEpicness commented:
ready for merge

nauticall reviewed (approved):
Sprites are fine, at least the ones on this PR. If there'll be new sprites on the CM xeno side as well they'll need another looking at

stanalbatross commented on code/datums/effects/xeno_strains/xeno_speed.dm:
```diff
@@ -26,6 +26,14 @@
+	..()
+	var/mob/living/carbon/affected_mob = affected_atom
+	if(HAS_TRAIT(affected_mob, TRAIT_CHARGING))
+		to_chat(affected_mob, SPAN_XENOWARNING("Your bowels are evacuated as you charge!"))
```
better message please.

stanalbatross commented on code/datums/effects/xeno_strains/xeno_speed.dm:
```diff
@@ -26,6 +26,14 @@
+		to_chat(affected_mob, SPAN_XENOWARNING("Your bowels are evacuated as you charge!"))
+		qdel(src)
+		return FALSE
+	return ..()
```
newline below

Carnage068 commented:
Welp its back now, we did it.

Wintermote commented:
> It does overall damage instead of targeted damage. This prevents getting organ damage/fracs on one charge or stupid insta delimbs.
> 
> Unless you want the latter, this is how it's going to be.

Still seems very weird. You're being hit by a blunt object, so getting bleeds on every body part just doesn't thematically fit. Like I'd expect that from a ravager ability, not the crusher.

PhantomEpicness commented:
> > It does overall damage instead of targeted damage. This prevents getting organ damage/fracs on one charge or stupid insta delimbs.
> > Unless you want the latter, this is how it's going to be.
> 
> Still seems very weird. You're being hit by a blunt object, so getting bleeds on every body part just doesn't thematically fit. Like I'd expect that from a ravager ability, not the crusher.
![Screenshot_419](https://user-images.githubusercontent.com/42422230/187832508-14555ce1-ccfd-4779-b555-e1d4656b1ade.png)

I looked deeper and brute damage over 10 has a random chance of causing bleeding. This is a product of human damage code and reworking damage is not the scope of this PR.

What we can do however, which I will note for the eventual balance patch, is lower the limb spread chance. This will focus the damage on fewer limbs, causing less overall damage but more focused damage, which I have tried to avoid because one hit massive organ damage is not fun.

Carnage068 commented:
> What we can do however, which I will note for the eventual balance patch, is lower the limb spread chance. This will focus the damage on fewer limbs, causing less overall damage, and therefore less overall bleeding, but more focused damage, which I have tried to avoid because one hit massive organ damage is not fun.

Or you can just lower the charge damage a bit overall. Would probably help, considering that some people are getting one shot without being hit by a full charge, and sometimes one shot without even getting hit by charge at all.

PhantomEpicness commented:
> > What we can do however, which I will note for the eventual balance patch, is lower the limb spread chance. This will focus the damage on fewer limbs, causing less overall damage, and therefore less overall bleeding, but more focused damage, which I have tried to avoid because one hit massive organ damage is not fun.
> 
> Or you can just lower the charge damage a bit overall. Would probably help, considering that some people are getting one shot without being hit by a full charge, and sometimes one shot without even getting hit by charge at all.

agreed

c4xmaniac commented:
Consider making the charger crusher faster, at least getting up to full momentum. It really takes a long time, and its really easy to stop the crusher from getting up to to charge speed. Its on the cusp of being fun/good, but its not there yet imo.

Carnage068 commented:
> Consider making the charger crusher faster, at least getting up to full momentum. It really takes a long time, and its really easy to stop the crusher from getting up to to charge speed. Its on the cusp of being fun/good, but its not there yet imo.

It's supposed to need room to start up a full charge. Being able to bodyblock it to prevent it from getting to that speed is just called good counterplay. At the same time, it shouldn't be almost completely immobile when it IS bodyblocked though. It would just make things too binary. One of the differences between Charger and Base Crusher is that it's supposed to be faster, not just sometimes, but in general. Or at least, that's what was fun for me. Being fast is just fun.
